src/librustc_trans/trans/intrinsic.rs

```diff
@@ -1270,3 +1323,335 @@ fn get_rust_try_fn<'a, 'tcx>(fcx: &FunctionContext<'a, 'tcx>,
+        let indices = match indices {
+            Some(i) => i,
+            None => return C_null(llret_ty)
+        };
+
+        return ShuffleVector(bcx, llargs[0], llargs[1], C_vector(&indices))
+    }
+
+    if name == "simd_insert" {
+        require!(arg_tys[0].is_simd(),
+                 "SIMD insert intrinsic monomorphized for non-SIMD input type");
+
+        let elem_ty = arg_tys[0].simd_type(tcx);
+        require!(arg_tys[2] == elem_ty,
+                 "SIMD insert intrinsic monomorphized with inserted type not SIMD element type");
+        return InsertElement(bcx, llargs[0], llargs[2], llargs[1])
+    }
+    if name == "simd_extract" {
+        require!(arg_tys[0].is_simd(),
+                 "SIMD insert intrinsic monomorphized for non-SIMD input type");
+
+        let elem_ty = arg_tys[0].simd_type(tcx);
+        require!(ret_ty == elem_ty,
+                 "SIMD insert intrinsic monomorphized with returned type not SIMD element type");
+        return ExtractElement(bcx, llargs[0], llargs[1])
+    }
+
+    if name == "simd_cast" {
+        require!(arg_tys[0].is_simd(),
+                 "SIMD cast intrinsic monomorphized with non-SIMD input type `{}`",
+                 arg_tys[0]);
+        require!(ret_ty.is_simd(),
+                 "SIMD cast intrinsic monomorphized with non-SIMD return type `{}`",
+                 ret_ty);
+        require!(arg_tys[0].simd_size(tcx) == ret_ty.simd_size(tcx),
+                 "SIMD cast intrinsic monomorphized with input type `{}` and \
+                  return type `{}` with different lengths: {} vs. {}",
+                 arg_tys[0],
+                 ret_ty,
+                 arg_tys[0].simd_size(tcx),
+                 ret_ty.simd_size(tcx));
+        // casting cares about nominal type, not just structural type
+        let in_ = arg_tys[0].simd_type(tcx);
+        let out = ret_ty.simd_type(tcx);
+
+        if in_ == out { return llargs[0]; }
+
+        match (&in_.sty, &out.sty) {
+            (&ty::TyInt(lhs), &ty::TyInt(rhs)) => {
+                match (lhs, rhs) {
+                    (ast::TyI8, ast::TyI8) |
+                    (ast::TyI16, ast::TyI16) |
+                    (ast::TyI32, ast::TyI32) |
+                    (ast::TyI64, ast::TyI64) => return llargs[0],
+
+                    (ast::TyI8, ast::TyI16) |
+                    (ast::TyI8, ast::TyI32) |
+                    (ast::TyI8, ast::TyI64) |
+                    (ast::TyI16, ast::TyI32) |
+                    (ast::TyI16, ast::TyI64) |
+                    (ast::TyI32, ast::TyI64) => return SExt(bcx, llargs[0], llret_ty),
+
+                    (ast::TyI16, ast::TyI8) |
+                    (ast::TyI32, ast::TyI8) |
+                    (ast::TyI32, ast::TyI16) |
+                    (ast::TyI64, ast::TyI8) |
+                    (ast::TyI64, ast::TyI16) |
+                    (ast::TyI64, ast::TyI32) => return Trunc(bcx, llargs[0], llret_ty),
+                    _ => {}
+                }
```

Comment:
Could these match blocks be consolidated a bit? It looks like they're all following a pretty common pattern and it'd help to tighten this up to refactor out the matches